doc/developer/design/20231113_optimizer_notice_catalog.md

```diff
@@ -0,0 +1,480 @@
+1. [Catalog Schema](#catalog-schema) section describes the design of a catalog
+   schema that contains the optimizer notices corresponding to the current
+   catalog state.
+2. [Schema Maintenance](#schema-maintenance) elaborates how the contents of the
+   proposed catalog schema will be maintained at runtime, including the
+   mechanisms mechanisms used to align the (externally visible) status of a
+   notice with the internal system state.
+
+### Console UX and Workflows
+
+This section contains wireframe diagrams that illustrate the console UX and
+workflows that will be supported by the proposed notices schema. Each wireframe
+links to [an Excalidraw scene][excalidraw_sources] which contains the wireframe
+sources and supports leaving comments directly on the diagram.
+
+#### Listing all Notices
+
+![Listing all objects wireframe](./static/20231113_optimizer_notice_catalog/ux_listing_all_notices.png)
+
+- For the MVP, all notices will be listed in a single page. In the future, we
+  can have scoped versions of this view as a tab on the clusters page.
+- The customer will be able to change the status of a notice directly from the
+  list view:
+  - A notice marked as "resolved" indicates that the customer has performed the
+    suggested corrective correction. This does not imply that the notice is
+    already resolved, but that we expect this to happen during the next once the
+    related catalog objects have been (re-)created.
+  - A notice marked as "ignored" indicates that the user considers it irrelevant
+    or low priority and they don't want to see it the foreseeable future.
+- The page will contain an extra filter dropdown that will allow customers to
+  filter out notices with a certain status:
+  - hide ignored and resolved - probably this should be the default,
+  - hide ignored - probably most useful for field engineers using the
+    impersonated console,
+  - show all.
+
+#### Inspecting a Notice
+
+![Listing all objects wireframe](./static/20231113_optimizer_notice_catalog/ux_inspecting_a_notice.png)
+
+- The _Notice_ and the _Hint_ sections correspond to what is currently rendered
+  in the `EXPLAIN` output. In addition, the detailed view contains a _Corrective
+  Action_ statement which contains concrete instructions that will resolve the
+  issue. If possible, the corrective action is given as a concrete SQL snippet,
+  otherwise it is a valid markdown string.
+- Every detailed view has a "Learn more" link below the notice section that
+  references the public docs for this notice type. The notice docs are
+  structured similar to [the Rust error codes page][rust_error_codes] and will
+  provide more context, a minimal reproducing example, and a corrective action
+  for the example for each notice type for customers that want to understand
+  where and why a specific notice is emitted.
+
+### Catalog Schema
+
+#### <a id="mz_optimizer_notices"></a> `mz_internal.mz_optimizer_notices` (new)
```

Comment:
I've been thinking that maybe instead of "Optimizer notices" we should call these things "Optimization notices", because some of them are actually emitted not by the optimizer. For example, `DroppedInUseIndex` is created in `sequence_drop_common`.

Reply:
I don't think that we are looking to maintain these notices in the same table. My impression was that the current iteration focuses precisely on the notices emitted by the optimizer as well as other notices that we can add in the future.

For example, dropping an index that is still in use will create dangling `import_ids` in some of the plans of running indexes, so we can have other notices attached to the specific index or materialized view that depends on that index.

Reply:
I will rename this to "Optimization notices".